pkg/fqdn/doc.go

```diff
@@ -12,20 +12,238 @@
+// consistent if the update order is different and merging multiple caches
+// should be equivalent to applying the constituent updates individually. As a
+// result, DNS data is all inserted into a single global cache from which the
+// policy layer receives information. This is historic and per-Endpoint
+// handling can be added. The data is internally tracked per IP because
+// overlapping DNS responses may have different TTLs for IPs that appear in
+// both.
+// Notes:
+// - The default configurable minimum TTL in the caches is 1 hour. This is
+//   mostly for identity stability, as short TTLs would cause more identity
+//   churn. This is mostly history as CIDR identities now have a near-0
+//   allocation overhead.
+// - DNSCache deletes only currently occur when the cilium API clears the cache
+//   or when the garbage collector evicts entries.
+// - The combination of caches: per-Endpoint and global must manage disparate
+//   behaviours of pods. The worst case scenario is one where one pod makes many
+//   requests to a target with changing IPs (like S3) but another makes few
+//   requests that are long-lived. We need to ensure "fairness" where one does
+//   not starve the other. The limits in the per-Endpoint caches allow this, and
+//   the global cache acts as a collector across different Endpoints (without
+//   restrictions).
+//
+// Expiration of DNS data is handled by the dns-garbage-collector-job controller.
+// Historically, the only expiration was TTL based and the per-Endpoint and
+// global caches would expire data at the same time without added logic.
+// This is not true when we apply per-host IP limits in the cache. These
+// default to 50 IPs for a given domain, per Endpoint. To account for these
+// evictions the controller handles TTL and IP limit evictions. This ensures
+// that the global cache is consistent with the per-Endpoint caches. The result
+// is that the actual expiration is imprecise (TTL especially). The caches mark
+// to-evict data internally and only do so on GC method calls from the
+// controller.
+// When DNS data is evicted from any per-Endpoint cache, for any reason, each
+// IP is retained as a "zombie". These are tracked per IP for that endpoint
+// (with a limit of 10000 by default). When the Connection Tracking garbage
+// collector runs, it marks any zombie IP that correlates to a live connection
+// by that endpoint as "alive". At the next iteration of the
+// dns-garbage-collector-job controller, the not-live zombies are finally
+// evicted. These IPs are then, finally, no longer placed into the global cache
+// on behalf of this endpoint. Other endpoints may have live DNS TTLs or
+// connections to the same IPs, however, so these IPs may be inserted into the
+// global cache for the same domain or a different one (or both).
+// Note: The CT GC has a variable run period. This ranges from 30s to 12 hours
+// and is shorter when more connection churn is observed.
```

Comment:
minor nit for maintainability, I wonder for these constants like 1000, 30s, 12h whether we should refer to the variable that holds these configurations?

Reply:
I was thinking about someone editing the default value, then wanting to grep to find references, so having the variable there should help keeping these docs up to date in such a case (hopefully).